rancher/rancher
# Add status field for global roles

## Issue: https://github.com/rancher/rancher/issues/42217
 
## Problem
When attempting to automate GlobalRoles, there is currently no method of knowing when the GlobalRole has successfully created all the backing Roles and ClusterRole. If an immediate next step was to check for or use the backing resources, it might fail. The only way around it would be to perform a wait and hoping it was adequately long enough. Similarly, updating a GlobalRole might require changes to the backing resources and there's no way to identify if those changes have taken place.

## Solution
This adds a new `Status` field to GlobalRoles. It contains the following fields:
- `Conditions`: essentially a status for each backing resource. If there are any errors attempting to create the backing resource they are specified here.
- `LastUpdate`: timestamp of the last update that changed the GlobalRole
- `Summary`: A string to indicate the overall status of the GlobalRole. Can be `InProgress`, `Complete` or `Error`

Any time a GlobalRole is created or updated, the status summary is set to `InProgress` until all the conditions (backing resources) are checked. Then, if all conditions are successful, the summary is set to `Complete`. If any condition is not met, it is set to `Error`.

Additionally, this PR adds much more unit test coverage of GlobalRoles.

## Testing


## Engineering Testing
### Manual Testing
On startup there are many default GlobalRoles, and those all get statuses set. I also added and modified my own GlobalRoles to test that the logic for InProgress, Complete and Error was correct. Another important test was to ensure that updating Namespaces or Roles would cause the reconcile to trigger. Finally, I tested to make sure the reconcile was not constantly being called. I wanted to limit the performance impact, which meant optimizing for the fewest reconciles possible.

### Automated Testing
* Test types added/modified:
    * Unit
Summary: There didn't exist unit tests to cover the `reconcileGlobalRole` and `reconcileCatalogRole` functions. I added full test coverage since I was modifying those functions.

## QA Testing Considerations
On a fresh install, the Status field is added to all the GlobalRoles that get created by default. In the case of an upgrade, they won't get a status field until they are modified in some way since the reconciler needs to run. 

Seeing the `InProgress` status may be difficult as the reconcile happens very quickly. The test can be done either by automation that calls "get" on the global role and watches the status. To do so manually, you can do `kubectl get globalrole <global-role-name> -w -oyaml` which will show the updates as they happen. 
 
### Regressions Considerations
This is a new field that is meant to be informative and not affect the logic of GlobalRoles. It is unlikely it would be responsible for regressions. The only concern may be performance related as it does add some more processing for each GlobalRole.

Existing / newly added automated tests that provide evidence there are no regressions:
Unfortunately this affects previously untested functions. The new unit tests will help identify regressions going forward.
